<h1>Capstone Project - The Battle of Neighborhoods (Week 4 and 5)</h1>

<h2>Introduction</h2> 
<p>For the final project, I would like to pizza restaurants in New York.</p>

<h2>Business Problem Section</h2> 
<h3>Background</h3>
<p>Luigi is the CEO of Luigi's Pizza Joint, which started as a family owned and operated business in the 60's. Luigi already has a number of pizza restaurants in other cities in America, but he would like to expand their gourmet pizza business in New York, specifically Manhattan.</p> 
<h3>Business Problem</h3>
<p>Luigi is concerned that there are already a number of pizza businesses in New York, specifically Manhattan, however none are to the same standard has Luigi's Pizza Joint. In this scenario, the aim is to identify a suitable location, on what basis is this choice suitable and identify the competition in that location.</p>
<p>The audience for the results will be Luigi and his five brothers who are also part directors of other Luigi's Pizza Joints across America.</p>

<h2>Data Section</h2> 
<h3>Data Sources</h3>
<p>Before choosing this topic I did try and locate data for Perth, Western Australia, however nothing free is available. Therefore I have chosen to use the New York data from the exercises (https://cocl.us/new_york_dataset).</p>
<p>I will also be using my personal developer Foursquare to locate other Pizza restaurants.</p> 
<h3>How the data will be used</h3>
<p>The main components of our analysis and predication system, this will comprise of four stages:
    
1. Collect Data
2. Explore and Understand Data
3. Data preparation and preprocessing 
4. Modeling
</p>

<h2>Collect Data</h2> 

In [3]:
## Install packages
!pip install bs4
!pip install lxml
!conda install -c conda-forge geopy --yes

     |████████████████████████████████| 102kB 10.2MB/s ta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/a0/b0/b2/4f80b9456b87abedbc0bf2d52235414c3467d8889be38dd472
Successfully built bs4
     |████████████████████████████████| 5.8MB 4.6MB/s eta 0:00:01�████████████████████           | 3.8MB 4.6MB/s eta 0:00:01     |████████████████████████████▏   | 5.1MB 4.6MB/s eta 0:00:01
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.11

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.49         |             py_0  

In [4]:
## Import libraries

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [5]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [6]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)


In [7]:
neighborhoods_data = newyork_data['features']

# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

## load data into the dataframe
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)
    
neighborhoods

Borough               Neighborhood   Latitude  Longitude
0            Bronx                  Wakefield  40.894705 -73.847201
1            Bronx                 Co-op City  40.874294 -73.829939
2            Bronx                Eastchester  40.887556 -73.827806
3            Bronx                  Fieldston  40.895437 -73.905643
4            Bronx                  Riverdale  40.890834 -73.912585
5            Bronx                Kingsbridge  40.881687 -73.902818
6        Manhattan                Marble Hill  40.876551 -73.910660
7            Bronx                   Woodlawn  40.898273 -73.867315
8            Bronx                    Norwood  40.877224 -73.879391
9            Bronx             Williamsbridge  40.881039 -73.857446
10           Bronx                 Baychester  40.866858 -73.835798
11           Bronx             Pelham Parkway  40.857413 -73.854756
12           Bronx                City Island  40.847247 -73.786488
13           Bronx               Bedford Park  40.870185 -73.885512
14           Bronx         University Heights  40.855727 -73.910416
15           Bronx             Morris Heights  40.847898 -73.919672
16           Bronx                    Fordham  40.860997 -73.896427
17           Bronx               East Tremont  40.842696 -73.887356
18           Bronx                 West Farms  40.839475 -73.877745
19           Bronx               High  Bridge  40.836623 -73.926102
20           Bronx                    Melrose  40.819754 -73.909422
21           Bronx                 Mott Haven  40.806239 -73.916100
22           Bronx                Port Morris  40.801664 -73.913221
23           Bronx                   Longwood  40.815099 -73.895788
24           Bronx                Hunts Point  40.809730 -73.883315
25           Bronx                 Morrisania  40.823592 -73.901506
26           Bronx                  Soundview  40.821012 -73.865746
27           Bronx               Clason Point  40.806551 -73.854144
28           Bronx                Throgs Neck  40.815109 -73.816350
29           Bronx               Country Club  40.844246 -73.824099
30           Bronx                Parkchester  40.837938 -73.856003
31           Bronx         Westchester Square  40.840619 -73.842194
32           Bronx                   Van Nest  40.843608 -73.866299
33           Bronx                Morris Park  40.847549 -73.850402
34           Bronx                    Belmont  40.857277 -73.888452
35           Bronx             Spuyten Duyvil  40.881395 -73.917190
36           Bronx            North Riverdale  40.908543 -73.904531
37           Bronx                 Pelham Bay  40.850641 -73.832074
38           Bronx              Schuylerville  40.826580 -73.826203
39           Bronx             Edgewater Park  40.821986 -73.813885
40           Bronx                Castle Hill  40.819014 -73.848027
41           Bronx                  Olinville  40.871371 -73.863324
42           Bronx             Pelham Gardens  40.862966 -73.841612
43           Bronx                  Concourse  40.834284 -73.915589
44           Bronx                  Unionport  40.829774 -73.850535
45           Bronx                   Edenwald  40.884561 -73.848083
46        Brooklyn                  Bay Ridge  40.625801 -74.030621
47        Brooklyn                Bensonhurst  40.611009 -73.995180
48        Brooklyn                Sunset Park  40.645103 -74.010316
49        Brooklyn                 Greenpoint  40.730201 -73.954241
50        Brooklyn                  Gravesend  40.595260 -73.973471
51        Brooklyn             Brighton Beach  40.576825 -73.965094
52        Brooklyn             Sheepshead Bay  40.586890 -73.943186
53        Brooklyn          Manhattan Terrace  40.614433 -73.957438
54        Brooklyn                   Flatbush  40.636326 -73.958401
55        Brooklyn              Crown Heights  40.670829 -73.943291
56        Brooklyn              East Flatbush  40.641718 -73.936103
57        Brooklyn                 Kensington  40.642382 -73.980421

In [8]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


In [9]:
# Creating new Dataframe manhattan_data
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [10]:
manhattan_data.shape[0]

40

In [11]:
# Define Foursquare Credentials and Version
CLIENT_ID = 'CLV4Q1JFT3JFZNHXENXHTTAHDZW0XIXOS5FRN3Z3X3NHFGPC' # your Foursquare ID
CLIENT_SECRET = 'QL0ZDQ24UUK4ZS4OTWZKYNNNAT0QXNRS4WKVA3I3WUNAP0VR' # your Foursquare Secret
VERSION = '20180604'
limit = 20
print('Your credentails:')
print('CLIENT_ID:'+ CLIENT_ID)
print('CLIENT_SECRET:'+ CLIENT_SECRET)

Your credentails:
CLIENT_ID:CLV4Q1JFT3JFZNHXENXHTTAHDZW0XIXOS5FRN3Z3X3NHFGPC
CLIENT_SECRET:QL0ZDQ24UUK4ZS4OTWZKYNNNAT0QXNRS4WKVA3I3WUNAP0VR


In [12]:
# Let's get the geographical coordinates of Manhattan.
address = 'Manhattan, NY'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:4: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  after removing the cwd from sys.path.


The geograpical coordinate of Manhattan are 40.7900869, -73.9598295.


In [13]:
# let's visualizat Manhattan the neighborhoods in it.
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

In [17]:
# Let's create a function to repeat the process to all the neighborhoods in New York
def getNearbyVenues(names, latitudes,longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names,latitudes,longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [18]:
# Write the code to run the above function on each neighborhood and create a new dataframe called venues.
venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude'],
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [19]:
print(venues.shape)
venues.head()

(4813, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896649       -73.844846   
2                         Carvel Ice Cream       40.890487       -73.848568   
3  Cooler Runnings Jamaican Restaurant Inc       40.898276       -73.850381   
4                                  Dunkin'       40.890459       -73.849089   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2        Ice Cream Shop  
3  Caribbean Restaurant  
4            Donut Shop

In [22]:
venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                               
Allerton                                      20                      20   
Annadale                                      12                      12   
Arden Heights                                  4                       4   
Arlington                                      6                       6   
Arrochar                                      18                      18   
Arverne                                       18                      18   
Astoria                                       20                      20   
Astoria Heights                               11                      11   
Auburndale                                    18                      18   
Bath Beach                                    20                      20   
Battery Park City                             20                      20   
Bay Ridge                                     20                      20   
Bay Terrace                                   32                      32   
Baychester                                    20                      20   
Bayside                                       20                      20   
Bayswater                                      2                       2   
Bedford Park                                  20                      20   
Bedford Stuyvesant                            20                      20   
Beechhurst                                    16                      16   
Bellaire                                      12                      12   
Belle Harbor                                  18                      18   
Bellerose                                     20                      20   
Belmont                                       20                      20   
Bensonhurst                                   20                      20   
Bergen Beach                                   7                       7   
Blissville                                    20                      20   
Bloomfield                                     5                       5   
Boerum Hill                                   20                      20   
Borough Park                                  20                      20   
Breezy Point                                   6                       6   
Briarwood                                     11                      11   
Brighton Beach                                20                      20   
Broad Channel                                  6                       6   
Broadway Junction                             14                      14   
Bronxdale                                     14                      14   
Brooklyn Heights                              20                      20   
Brookville                                     1                       1   
Brownsville                                   20                      20   
Bulls Head                                    20                      20   
Bushwick                                      20                      20   
Butler Manor                                   7                       7   
Cambria Heights                               12                      12   
Canarsie                                       6                       6   
Carnegie Hill                                 20                      20   
Carroll Gardens                               20                      20   
Castle Hill                                    7                       7   
Castleton Corners                             15                      15   
Central Harlem                                20                      20   
Charleston                                    20                      20   
Chelsea                                       25                      25   
Chinatown                                     20                      20   

In [23]:
# Let's find out how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(venues['Venue Category'].unique())))

There are 349 uniques categories.


<h2>Analyse Data</h2> 

In [24]:
# one hot encoding
onehot = pd.get_dummies(venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
onehot['Neighborhood'] = venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [onehot.columns[-1]] + list(onehot.columns[:-1])
onehot = onehot[fixed_columns]

onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0       0                 0                       0            0           0   
1       0                 0                       0            0           0   
2       0                 0                       0            0           0   
3       0                 0                       0            0           0   
4       0                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Automotive Shop  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       0     0    0               0                 0   

   Basketball Court  Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  \
0                 0      0          0                0         0            0   
1                 0      0          0                0         0            0   
2                 0      0          0                0         0            0   
3                 0      0          0                0         0            0   
4                 0      0          0                0         0            0   

   Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Shop  Bike Trail  \
0           0              0                         0          0           0   
1           0              0                         0          0           0   
2           0              0                         0          0           0   
3           0              0                         0          0           0   
4           0              0                         0          0           0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  Bowling Alley  \
0       0           0              0          0         0              0   
1       0           0              0          0         0              0   
2       0           0              0          0         0              0   
3       0           0              0          0         0              0   
4       0           0              0          0         0              0   

   Boxing Gym  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0           0                     0               0        0            0   
1           0                     0               0        0 

In [26]:
# Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
grouped = onehot.groupby('Neighborhood').mean().reset_index()

In [27]:
# Let's print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = grouped[grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')


----Allerton----
               venue  freq
0        Pizza Place  0.15
1           Pharmacy  0.10
2        Supermarket  0.10
3        Bus Station  0.05
4  Martial Arts Dojo  0.05


----Annadale----
                 venue  freq
0  American Restaurant  0.17
1           Sports Bar  0.08
2           Restaurant  0.08
3                 Food  0.08
4        Train Station  0.08


----Arden Heights----
           venue  freq
0    Pizza Place  0.25
1  Deli / Bodega  0.25
2       Pharmacy  0.25
3    Coffee Shop  0.25
4   Optical Shop  0.00


----Arlington----
                 venue  freq
0        Deli / Bodega  0.17
1  American Restaurant  0.17
2          Coffee Shop  0.17
3             Bus Stop  0.17
4        Boat or Ferry  0.17


----Arrochar----
                venue  freq
0            Bus Stop  0.17
1       Deli / Bodega  0.11
2  Italian Restaurant  0.11
3         Pizza Place  0.06
4               Hotel  0.06


----Arverne----
            venue  freq
0       Surf Spot  0.22
1   Metro Station  

In [28]:
# Let's put that into a pandas dataframe
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
# Now let's create the new dataframe and display the top 10 venues for each neighborhood.
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = grouped['Neighborhood']

for ind in np.arange(grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood      1st Most Common Venue  \
0                     Allerton                Pizza Place   
1                     Annadale        American Restaurant   
2                Arden Heights                   Pharmacy   
3                    Arlington               Intersection   
4                     Arrochar                   Bus Stop   
5                      Arverne                  Surf Spot   
6                      Astoria         Seafood Restaurant   
7              Astoria Heights                Supermarket   
8                   Auburndale              Deli / Bodega   
9                   Bath Beach            Bubble Tea Shop   
10           Battery Park City                       Park   
11                   Bay Ridge                        Spa   
12                 Bay Terrace                Supermarket   
13                  Baychester                 Donut Shop   
14                     Bayside           Greek Restaurant   
15                   Bayswater                       Park   
16                Bedford Park                      Diner   
17          Bedford Stuyvesant                        Bar   
18                  Beechhurst                Yoga Studio   
19                    Bellaire              Deli / Bodega   
20                Belle Harbor                      Beach   
21                   Bellerose              Deli / Bodega   
22                     Belmont         Italian Restaurant   
23                 Bensonhurst             Ice Cream Shop   
24                Bergen Beach            Harbor / Marina   
25                  Blissville                      Hotel   
26                  Bloomfield                 Theme Park   
27                 Boerum Hill                        Spa   
28                Borough Park                       Bank   
29                Breezy Point        Monument / Landmark   
30                   Briarwood              Deli / Bodega   
31              Brighton Beach                 Restaurant   
32               Broad Channel              Deli / Bodega   
33           Broadway Junction              Deli / Bodega   
34                   Bronxdale         Italian Restaurant   
35            Brooklyn Heights                Yoga Studio   
36                  Brookville              Deli / Bodega   
37                 Brownsville                 Restaurant   
38                  Bulls Head                   Pharmacy   
39                    Bushwick                Coffee Shop   
40                Butler Manor             Baseball Field   
41             Cambria Heights       Caribbean Restaurant   
42                    Canarsie                        Gym   
43               Carnegie Hill       Gym / Fitness Center   
44             Carroll Gardens                Coffee Shop   
45                 Castle Hill              Deli / Bodega   
46           Castleton Corners                Pizza Place   
47              Central Harlem         African Restaurant   
48                  Charleston             Cosmetics Shop   
49                     Chelsea                    Theater   
50                   Chinatown         Chinese Restaurant   
51                 City Island     Thrift / Vintage Store   
52                   City Line                 Donut Shop   
53                Civic Center         Falafel Restaurant   
54           Claremont Village                Bus Station   
55                Clason Point                       Park   
56                     Clifton              Grocery Store   
57                     Clinton       Gym / Fitness Center   
58                Clinton Hill            Thai Restaurant   
59                  Co-op City       Fast Food Restaurant   
60                 Cobble Hill               Cocktail Bar   
61               College Point              Deli / Bodega   
62                     Concord       Gym / Fitness Center   
63                   Concourse                Pizza Place   
64           Concourse Village                   Phar

<h2>Clustering Manhattan</h2> 

In [57]:
# Run k-means to cluster the neighborhood into 5 clusters.
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 1, 0, 0, 0, 2, 1, 1, 1, 1], dtype=int32)

In [60]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
manhattan_merged = manhattan_data

# merge grouped with manhattan_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head(30) 

Borough         Neighborhood   Latitude  Longitude  Cluster Labels  \
0   Manhattan          Marble Hill  40.876551 -73.910660               2   
1   Manhattan            Chinatown  40.715618 -73.994279               2   
2   Manhattan   Washington Heights  40.851903 -73.936900               1   
3   Manhattan               Inwood  40.867684 -73.921210               1   
4   Manhattan     Hamilton Heights  40.823604 -73.949688               1   
5   Manhattan       Manhattanville  40.816934 -73.957385               1   
6   Manhattan       Central Harlem  40.815976 -73.943211               1   
7   Manhattan          East Harlem  40.792249 -73.944182               1   
8   Manhattan      Upper East Side  40.775639 -73.960508               1   
9   Manhattan            Yorkville  40.775930 -73.947118               1   
10  Manhattan           Lenox Hill  40.768113 -73.958860               1   
11  Manhattan     Roosevelt Island  40.762160 -73.949168               1   
12  Manhattan      Upper West Side  40.787658 -73.977059               1   
13  Manhattan       Lincoln Square  40.773529 -73.985338               1   
14  Manhattan              Clinton  40.759101 -73.996119               1   
15  Manhattan              Midtown  40.754691 -73.981669               1   
16  Manhattan          Murray Hill  40.748303 -73.978332               1   
17  Manhattan              Chelsea  40.744035 -74.003116               1   
18  Manhattan    Greenwich Village  40.726933 -73.999914               1   
19  Manhattan         East Village  40.727847 -73.982226               1   
20  Manhattan      Lower East Side  40.717807 -73.980890               1   
21  Manhattan              Tribeca  40.721522 -74.010683               1   
22  Manhattan         Little Italy  40.719324 -73.997305               1   
23  Manhattan                 Soho  40.722184 -74.000657               1   
24  Manhattan         West Village  40.734434 -74.006180               1   
25  Manhattan     Manhattan Valley  40.797307 -73.964286               1   
26  Manhattan  Morningside Heights  40.808000 -73.963896               1   
27  Manhattan             Gramercy  40.737210 -73.981376               2   
28  Manhattan    Battery Park City  40.711932 -74.016869               1   
29  Manhattan   Financial District  40.707107 -74.010665               1   

   1st Most Common Venue 2nd Most Common Venue          3rd Most Common Venue  \
0         Discount Store           Coffee Shop                    Yoga Studio   
1     Chinese Restaurant          Noodle House                 Sandwich Place   
2                   Park             Wine Shop                           Café   
3                   Café              Wine Bar                           Park   
4            Yoga Studio  Caribbean Restaurant             Mexican Restaurant   
5     Italian Restaurant       Other Nightlife      Japanese Curry Restaurant   
6     African Restaurant   American Restaurant                   Dessert Shop   
7     Mexican Restaurant                  Park                     Steakhouse   
8                  Hotel        Chocolate Shop                            Bar   
9              Wine Shop         Deli / Bodega                    Coffee Shop   
10                   Gym       Thai Restaurant                  Women's Store   
11        Sandwich Place                  Park                    Coffee Shop   
12    Italian Restaurant    Chinese Restaurant  Vegetarian / Vegan Restaurant   
13   Indie Movie Theater          Concert Hall                        Theater   
14  Gym / Fitness Center               Theater                  Indie Theater   
15                 Hotel           Salad Place             Miscellaneous Shop   
16     Korean Restaurant           Coffee Shop                          Hotel   
17               Theater              Bus Stop                      Nightclub   
18     French Restaurant    Italian Restaurant                           Café   
19            Beer Store     

In [66]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [64]:
#Commercial Places
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]


Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
2    Washington Heights                  Park             Wine Shop   
3                Inwood                  Café              Wine Bar   
4      Hamilton Heights           Yoga Studio  Caribbean Restaurant   
5        Manhattanville    Italian Restaurant       Other Nightlife   
6        Central Harlem    African Restaurant   American Restaurant   
7           East Harlem    Mexican Restaurant                  Park   
8       Upper East Side                 Hotel        Chocolate Shop   
9             Yorkville             Wine Shop         Deli / Bodega   
10           Lenox Hill                   Gym       Thai Restaurant   
11     Roosevelt Island        Sandwich Place                  Park   
12      Upper West Side    Italian Restaurant    Chinese Restaurant   
13       Lincoln Square   Indie Movie Theater          Concert Hall   
14              Clinton  Gym / Fitness Center               Theater   
15              Midtown                 Hotel           Salad Place   
16          Murray Hill     Korean Restaurant           Coffee Shop   
17              Chelsea               Theater              Bus Stop   
18    Greenwich Village     French Restaurant    Italian Restaurant   
19         East Village            Beer Store     Korean Restaurant   
20      Lower East Side           Art Gallery           Coffee Shop   
21              Tribeca             Wine Shop   American Restaurant   
22         Little Italy        Ice Cream Shop              Wine Bar   
23                 Soho         Women's Store        Clothing Store   
24         West Village    Italian Restaurant           Coffee Shop   
25     Manhattan Valley         Deli / Bodega                   Bar   
26  Morningside Heights             Bookstore   American Restaurant   
28    Battery Park City                  Park            Food Court   
29   Financial District         Jewelry Store  Gym / Fitness Center   
30        Carnegie Hill  Gym / Fitness Center    Italian Restaurant   
31                 Noho    Italian Restaurant     French Restaurant   
32         Civic Center    Falafel Restaurant                   Spa   
33        Midtown South     Korean Restaurant        Lingerie Store   
34         Sutton Place                   Gym     Indian Restaurant   
35           Turtle Bay      Sushi Restaurant           Karaoke Bar   
36           Tudor City                  Park           Yoga Studio   
37      Stuyvesant Town            Playground                   Bar   
38             Flatiron          Cycle Studio   Japanese Restaurant   
39         Hudson Yards   American Restaurant  Gym / Fitness Center   

            3rd Most Common Venue     4th Most Common Venue  \
2                            Café             Deli / Bodega   
3                            Park               Yoga Studio   
4              Mexican Restaurant              Cocktail Bar   
5       Japanese Curry Restaurant               Beer Garden   
6                    Dessert Shop                 Juice Bar   
7                      Steakhouse            Sandwich Place   
8                             Bar                    Bakery   
9                     Coffee Shop                       Gym   
10                  Women's Store               Pizza Place   
11                    Coffee Shop          Greek Restaurant   
12  Vegetarian / Vegan Restaurant                Bagel Shop   
13                        Theater     Performing Arts Venue   
14                  Indie Theater               Pizza Place   
15             Miscellaneous Shop  Mediterranean Restaurant   
16                          Hotel               Supermarket   
17                      Nightclub        Italian Restaurant   
18                           Café          Sushi Restaurant   
19                     Bagel Shop       Moroccan Restaurant   
20                  Women's Store                  Pet Café   
21                           Park             Indie 

In [65]:
#Tourist Areas & Hubs
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0   Marble Hill        Discount Store           Coffee Shop   
1     Chinatown    Chinese Restaurant          Noodle House   
27     Gramercy           Pizza Place           Coffee Shop   

   3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0            Yoga Studio                    Bank                   Gym   
1         Sandwich Place          Ice Cream Shop          Cocktail Bar   
27           Yoga Studio  Thrift / Vintage Store             Irish Pub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Donut Shop                 Diner              Pharmacy   
1                 Museum             Bike Shop         Garden Center   
27          Gourmet Shop          Liquor Store    Mexican Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0       Department Store            Pizza Place  
1               Tea Room     English Restaurant  
27            Food Truck             Playground

<h3>Output Findings to CSV</h3>

In [71]:
## Output Manhattan
manhattan_merged.to_csv('Manhattan_Merged.csv',index=False)

In [70]:
## Output all neighbourhoods for alternatives
neighborhoods_venues_sorted.to_csv('NewYork.csv',index=False)

<h3>Results</h3>
<p>From analysing the data, we have uncovered that Gramercy has the most common venue for pizza places in a tourist neighborhoods. Lenox Hill and Clinton has the fourth most common venue for pizza places and Marble Hill is tenth in the commercial neighborhoods.
    
If Luigi would like to open his newest restaurant in Manhattan, Gramercy has the most amount of pizza places / venues at 15% of all venues. It would be stiff competition, however Luigi's pizza's are proven winners in other cities.
</p>

<h3>Observations</h3>
<p>When analysing the data from other neighborhoods in New York, pizza places are generally quite popular in general. 33 out of the 301 neighborhoods in New York had a 1st Most Common Venue, so pizza places are definitely popular, however it will be competitive.</p>